### PR TITLE
[fix 3853] query mailserver when going back online

### LINKED ISSUE #3853: Can't receive offline inbox messages until re-login to account
*Type*: Bug
*Summary*: Messages sent to contact which is offline - are not received after contact appears online. However, after re-login to account - sometimes not from the first try, - but offline messages received and in the appropriate order.
**Note:** It was confirmed that offline messages are received (after several account re-logins) while both contacts on Mainnet with upstream RPC and Ropsten with upstream RPC.

#### Expected behavior
Messages are received from an offline inbox after contact appears online 

#### Actual behavior
Messages are not received from an offline inbox after contact appears online 

### Reproduction
1) Install fresh develop build on Device 1 and Device 2
2) Create Contact1 and Contact2 respectively
3) Both contacts are online and exchanged a text messages
4) Contact2 enables airplane mode on device
5) Contact1 sends 10 text messages to Contact2
6) Contact2 disables airplane mode and stays on Home/1-1chat screen for 5+ minutes
**Actual result:** No messages received.

7) Contact2 re-logins to the account (or stops and starts the app)
**Actual result:** Messages are received from offline inbox. **Note:** It may require repeating several times this step as not always it works.

### Additional Information
* Status version: Develop nightly build (336th in Jenkins)
* Operating System: Android and iOS

#### Logs

From TF session https://app.testfairy.com/projects/4803622-status/builds/8092250/sessions/36/?accessToken=d6n--lIHkdqqi2Qo-9-ovwyxejc
noticed in logs:
```00:55 I/ReactNativeJS: INFO [status-im.transport.inbox:154] - Peer enode://fa63a6cc730468c5456eab365b2a7a68a166845423c8c9acc363e5f8c4699ff6d954e7ec58f13ae49568600cff9899561b54f6fc2b9923136cd7104911f31cce@163.172.168.202:30303 is not registered. Retrying fetch peers.
00:55 E/ReactNativeJS: ERROR [status-im.transport.inbox:157] - Could not connect to mailserver
00:55 E/unknown:ReactNative: console.error: "ERROR [status-im.transport.inbox:157] - Could not connect to mailserver", stack:
cljs$core$IFn$_invoke$arity$3@12:347932
cljs$core$IFn$_invoke$arity$2@12:347729
cljs$core$IFn$_invoke$arity$2@12:353001
c@12:5508618
r@12:5508800
<unknown>@12:3240368
<unknown>@12:3259703
cljs$core$IKVReduce$_kv_reduce$arity$3@12:545631
_kv_reduce@12:158196
reduce_kv@12:268208
cljs$core$IFn$_invoke$arity$10@12:3257009
<unknown>@12:6012645
<unknown>@12:3950520
invoke_interceptor_fn@12:3878457
invoke_interceptors@12:3878758
execute@12:3880455
handle@12:3890180
re_frame$router$IEventQueue$_process_1st_event_in_queue$arity$1@12:3924913
re_frame$router$IEventQueue$_run_queue$arity$1@12:3905649
<unknown>@12:3911401
re_frame$router$IEventQueue$_fsm_trigger$arity$3@12:3915612
<unknown>@12:3908167
<unknown>@32:1369
x@32:643
callTimers@32:2788
value@27:3582
<unknown>@27:1067
value@27:3009
value@27:1039
02:27 D/ViewRootImpl: ViewPostImeInputStage ```

<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8092250/sessions/36/?accessToken=d6n--lIHkdqqi2Qo-9-ovwyxejc">TestFairy: Status 0.9.10-869-g7148734e+ (2426) @36</a></strong></div></blockquote>

### PR DESCRIPTION
**Pull request by yenda**
fix #3853

- also download history going back online (couldn't do it with from parameter, investigating with go-team)
- download history when joining public group chat

status: ready

### CODE REVIEW

**Comment by flexsurfer** [comment]
#3865 merged

**Review by janherich — APPROVED** [review]

**Review by cammellos — APPROVED** [review]

**Comment by asemiankevich** [comment]
@yenda Eric will you work on something else in terms of this PR (look into https://github.com/status-im/status-react/issues/3853#issuecomment-383080900) ? 

seems git hub bot moved this to testing again , but I understood that there is something we can do better in this PR prio merge, right?

File: src/status_im/transport/inbox.cljs [hunk 1]
```diff
@@ -149,44 +180,52 @@
...
+          from     (or from (:inbox/last-request db) nil)
+          sym-key-id (:inbox/sym-key-id db)]
       {::request-messages {:wnode      wnode
                            :topics     topics
-                           :to         to
                            :from       from
                            :sym-key-id sym-key-id
-                           :web3       web3}})))
+                           :web3       web3}
+       :db (assoc db :inbox/last-request (quot now 1000))})))
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs]: What if the request fails? Should we update db only on successful fetch?

Comment (cammellos) [reply src/status_im/transport/inbox.cljs]: Also maybe worth adding a comment addressing what is the behaviour if our clock is not synchronized,  at a first glance we could be missing some messages:

Our clock is +10s over the mailserver, assume MS is perfectly synchronized (0). 
we ask for 0 and set last-request to 10s
a message arrives on the mailserver with timestamp 5
next time we ask for 10, missing 5.

Is this possible? (don't know much about how mailserver handles queries)

File: src/status_im/transport/inbox.cljs [hunk 2]
```diff
@@ -133,14 +157,21 @@
 ;;;; Handlers
 
 (handlers/register-handler-fx
-  ::fetch-peers
+  :inbox/add-peer
+  ;; This event adds a wnode to the list of peers
+  (fn [_ [_ wnode]]
+    (println ::add-peer)
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs]: println :)

File: src/status_im/transport/inbox.cljs (hunk 1 above)
Comment (yenda) [reply src/status_im/transport/inbox.cljs]: Interesting question, I don't know if it really matters, anyway I had to drop the from field after more testing it is quite obvious that the mailserver is dropping the peer when using the from field. So I will leave that for now, wait for the logs tomorrow and see what can be done. I think this PR is good to go without it for now.

**Review by cammellos — APPROVED** [review]
@yenda We also had a chat at some point some time ago with the test team, the desired behaviour is that we should not be fetching history from public chats (so `from` should be initialized with the joining time). It's worth you double check that with them.
In this case we are not using `from` so it does not matter anyway at the moment. 

Otherwise the PR looks ok to me, maybe worth leaving a comment so we know where we want to be? (We had quite a few recursive discussion already, so a comment would put the last nail in the coffin hopefully). 

Also you still have a println :)

**Comment by janherich** [comment]
@yenda Please update the original description of the PR, now that `from` field for `requesteMessages` method had to be dropped.
Also I'm little bit uneasy about `async-flow-tx` library, as I believe we could do something more fitting our current direction ourselves (I will prepare gist of solution today), without relying on external dep, but that's not so important point now.

**Review by janherich — APPROVED** [review]

**Comment by yenda** [comment]
@janherich looking forward to your proposal but I am not really concerned about this dependency it is from the same company that does re-frame and it is pretty sound, nice DSL to describe async flows.

**Comment by Serhy** [comment]
Builds Android: https://i.diawi.com/KXGUzj and iOS: https://i.diawi.com/yakoSj has been tested. Everything here works as expected: once contact appears online the message history pulls out from offline inbox.

I've also measured how battery drains here comparing to develop and results are equal: 
So for the next scenario:
1) Contact1 sends to Contact2 100 messages in 1-1 chat, then 100 messages in GroupChat and 100 messages in public chat.
2) Contact2 turns on airplane mode.
3) Contact1 sends to Contact2 20 messages in 1-1 chat, then 20 messages in GroupChat and 20 messages in public chat.
4) Contact2 turns off airplane mode and stays on Home screen to receive messages sent while he was offline.

Device power use in both cases is 0.41% for im.status.ethereum. 

Ready for merge! 
 

<blockquote><div><strong><a href="https://i.diawi.com/KXGUzj">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>
<blockquote><div><strong><a href="https://i.diawi.com/yakoSj">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Pull request merged**